boilermaker/evaluators/common.py

```diff
@@ -0,0 +1,393 @@
+            )
+            return None
+
+        if task.should_dead_letter:
+            await cls.dead_letter_message(
+                task.msg,
+                receiver,
+                reason=reason,
+                error_description=detail or "Task failed",
+            )
+        else:
+            await cls.complete_message(task.msg, receiver)
+        return None
+
+
+class MessageHandler:
+    """
+    Handles messages from the service bus.
+    This is a base class to be extended by specific evaluators.
+    It provides common functionality for handling messages, including
+    decoding tasks, managing message locks, and publishing tasks.
+    """
+
+    def __init__(
+        self,
+        receiver: ServiceBusReceiver,
+        task: Task,
+        task_publisher: Callable[[Task], Awaitable[None]],
+        function_registry: dict[str, TaskHandler],
+        state: typing.Any | None = None,
+        storage_interface: StorageInterface | None = None,
+    ):
+        self._receiver: ServiceBusReceiver = receiver
+        self.task = task
+        self.function_registry = function_registry
+        self.state = state if state is not None else {}
+        self.task_publisher = task_publisher
+        self.storage_interface = storage_interface
+
+    async def __call__(self):
+        return await self.message_handler()
+
+    # TO BE IMPLEMENTED BY SUBCLASSES
+    @abstractmethod
+    async def message_handler(self):
+        """Individual message handler"""
+        raise NotImplementedError
+
+    # Shared properties and methods
+    @cached_property
+    def current_msg(self) -> ServiceBusReceivedMessage | None:
+        return self.task.msg
+
+    @cached_property
+    def sequence_number(self) -> int | None:
+        return self.task.sequence_number
+
+    async def publish_task(
+        self,
+        task: Task,
+        delay: int = 0,
+        publish_attempts: int = 1,
+    ) -> Task:
+        await self.task_publisher(task, delay=delay, publish_attempts=publish_attempts)
+        return task
+
+    # Message handling actions
+    async def abandon_current_message(self) -> None:
+        return await MessageActions.abandon_message(self.current_msg, self._receiver)
+
+    async def complete_message(self) -> None:
+        return await MessageActions.complete_message(self.current_msg, self._receiver)
+
+    async def renew_message_lock(self) -> datetime.datetime | None:
+        """Renew the lock on the current message being processed."""
+        return await MessageActions.renew_message_lock(self.current_msg, self._receiver)
+
+    async def deadletter_or_complete_task(
+        self,
+        reason: str,
+        detail: Exception | str | None = None,
+    ) -> None:
+        return await MessageActions.deadletter_or_complete_task(
+            self.task, self._receiver, reason, detail=detail
+        )
+
+    async def task_handler(self) -> TaskResult:
```

Comment:
@add1 , two major changes (amidst the massive reworking here...):
1. Build a `task_handler` method that evaluates a task and returns a `TaskResult`, which indicates whether we succeeded, failed, or needed a retry, and
2. ...see simple `message_handler` below...